nesta_ds_utils/file_ops.py

```diff
@@ -1,53 +1,196 @@
+
+    Returns:
+        Path: Path in pathlib format.
     """
     return Path(path) if type(path) is str else path
 
 
 def make_path_if_not_exist(path: Union[Path, str]):
     """Check if path exists, if it does not exist then create it.
 
-    :param path: file path 
-    :type path: pathlib.Path or str
+    Args:
+        path (Union[Path, str]): File path.
     """
     path = _convert_str_to_pathlib_path(path)
     if not path.exists():
         path.mkdir(parents=True)
 
 
 def extractall(
-    zip_path: Union[Path, str], 
-    out_path: Union[Path, str]=None, 
-    delete_zip: Boolean = True):
+    zip_path: Union[Path, str],
+    out_path: Union[Path, str] = None,
+    delete_zip: Boolean = True,
+):
     """Takes path to zipped file and extracts it to specified output path.
-    
-    :param zip_path: path to zipped file
-    :type zip_path: str or pathlib.Path
-    :param out_path: path where contents will be unzipped to
-    :type out_path: str or pathlib.Path
-    :param delete_zip: option to delete zip file after extracted
-    :type delete_zip: Boolean
-    """
 
+    Args:
+        zip_path (Union[Path, str]): Path to zipped file.
+        out_path (Union[Path, str], optional): Path where contents will be unzipped to. Defaults to None.
+        delete_zip (Boolean, optional): Option to delete zip file after extracted. Defaults to True.
+    """
     if out_path is None:
         out_path = zip_path.rsplit("/", 1)[0]
 
     make_path_if_not_exist(out_path)
     with zipfile.ZipFile(zip_path, "r") as z:
         z.extractall(out_path)
-    
+
     if delete_zip is True:
-        os.remove(zip_path)
\ No newline at end of file
+        os.remove(zip_path)
+
+
+def get_bucket_filenames_s3(bucket_name: str, dir_name: str = "") -> List[str]:
+    """Get a list of all files in bucket directory.
+
+    Args:
+        bucket_name (str): S3 bucket name
+        dir_name (str, optional): Directory or sub-directory within an S3 bucket.
+            Defaults to '' (the top level bucket directory).
+
+    Returns:
+        List[str]: List of file names in bucket directory.
+    """
+    s3_resources = boto3.resource("s3")
+    my_bucket = s3_resources.Bucket(bucket_name)
+    return [
+        object_summary.key
+        for object_summary in my_bucket.objects.filter(Prefix=dir_name)
+    ]
+
+
+def _df_to_fileobj(df: pd.DataFrame, save_file_dir: str) -> io.BytesIO:
+    """Convert DataFrame into bytes file object.
+
+    Args:
+        df (pd.DataFrame): Dataframe to convert.
+        save_file_dir (str): Saving file name.
+
+    Returns:
+        io.BytesIO: Bytes file object.
+    """
+    buffer = io.BytesIO()
+    if fnmatch(save_file_dir, "*.csv"):
+        df.to_csv(buffer)
+    elif fnmatch(save_file_dir, "*.json"):
+        df.to_json(buffer)
+    elif fnmatch(save_file_dir, "*.pkl"):
+        df.to_pickle(buffer)
+    elif fnmatch(save_file_dir, "*.xml"):
+        df.to_xml(buffer)
```

Comment:
All the `.to_{something}` functions take other kwargs which might be important for the user such as whether to save the index or not. Can we make it so that the top level function `upload_obj` takes in a dictionary which can then be unpacked into these?

I also wonder whether we should completely restrict saving dataframes to csv and parquet formats. I think in the vast majority of cases anyone saving a dataframe will want these and anyone saving a json, xml etc. will probably be doing so from a dict, list of dicts or xml object etc.

Reply:
Ok, I can restrict restrict dataframe to csv and parquet and add kwargs for the pandas functions.